### ISSUE TITLE
Assignment 1 is now open

### ISSUE THREAD

**Issue by vincenzocoia** [vincenzocoia/BAIT509#5]

**Comment by vincenzocoia** [comment]
Update: You are now allowed to use either R or python for your assignment. You don't have to use Jupyter notebooks if you don't want to.

This change has been reflected in the assignment instructions -- you can see exactly what was changed [here](https://github.com/vincenzocoia/BAIT509/commit/ffc14457dfdf16b2f262d3084403b0281bc2fe13#diff-bae47f458e7ac7dcbff731e8feacc82f) (this is called the "diff" in git).

**Comment by vincenzocoia** [comment]
__Extension__: Assignment 1 is now due Monday, March 12 at 10:00am (instead of March 7). Changes are reflected throughout the repo.

**Comment by vincenzocoia** [comment]
UBC Connect should now be open! You'll be able to see the "Assignment 1" content, where you can submit your homework. Please let me know if you can't see this.

**Comment by Maggielena** [comment]
Hi Vincenzo:

While doing the Assignment 1, I came cross this confusion matrix of KNN. Is this something that we can use for our assignment?

Thank you!
Maggie

**Comment by vincenzocoia** [comment]
@Maggielena great to hear that you're looking more into kNN! Confusion matrices are a way of spelling out classification error in more detail, by indicating exactly where and when a mistake is made. It's not something we will be focussing on in this course, but definitely important if one wishes to dig deeper into classification problems. Nice find!

**Comment by ChangHe591** [comment]
Hello Vincenzo,

A quick question regarding loess. It says that "To include Sex in your model, simply fit a loess model to each of "male" and "female". " Do you mean we should have two loess models-one for female and one for male?

In addition, you mention that x could take any integer values between 0 to 15. How come when I view the data it also has a value of 16?

![image](https://user-images.githubusercontent.com/36861381/37181925-d6cacdaa-22e3-11e8-834d-ae01125e9ccf.png)
Much appreciated.

Thanks,
Chang

**Comment by vincenzocoia** [comment]
Hi @ChangHe591:

> Do you mean we should have two loess models-one for female and one for male?

Exactly.

> How come when I view the data it also has a value of 16?

Oops, must be between 0 and 16, then. The actual range doesn't matter though -- all you have to know is that the numbers represent lightness of a pixel.

**Comment by Maggielena** [comment]
Hi Vincenzo,

I am trying to apply the np.random.shuffle (X) function to variable X_new, but my X_new is always empty. I could not see why this is not working ... 

P.S.: In the assignment question, I think it is okay to not to shuffle it though.. 

Thank you!!

Maggie

**Comment by vincenzocoia** [comment]
@Maggielena :

> I am trying to apply the np.random.shuffle (X) function to variable X_new, but my X_new is always empty. I could not see why this is not working ...

Are you trying to shuffle data after training the model? If so, this is not necessary -- it's only useful _before_ splitting your data into training and test/validation sets. If this is not the case, and it's a coding problem, I can't quite tell where the error comes in -- a minimum working example would be useful. 

> P.S.: In the assignment question, I think it is okay to not to shuffle it though..

Probably. It's just good practice to do so. Also helps keep your analysis robust in case more/different data are used.

----------

Also, a student had asked a few questions by email that I'll address here.

> Are the discussion of scaling based on log.fare or original data?

Whichever you want. You'll find that your data are more "together" if you use log(fare) -- and I'd imagine this would result in a better fit. If I recall correctly, there are a few zeroes for the "fare" variable, which you'd end up losing with the log transformation, but these things happen... making with predictions when fare=0 isn't particularly important here, so you can ignore those cases.

> For the discussion about loess, do we set the degree at 1? or we will need to discuss how the change of degree affects the prediction?

I'm not quite sure what question is being referred to here.

> In addition, how do we deal with missing data? Do we just omit them?

For this assignment, you can just remove observations where one of the predictors or response is missing. In general though, there _are_ methods that have been developed to deal with missing data. The most widely accepted method is called _multiple imputation_. Not required for this course.

**Comment by Maggielena** [comment]
Hi Vincenzo,

My code is very simple.. 
X = digits['data']
X_new = np.random.shuffle (X)

I am probably missing some steps here...

Thank you

Maggie

**Comment by kumar-srikanth** [comment]
You actually don't need to shuffle the data.
If you look at the documentation of train_test_split function, it already
has a randomizer inside it which shuffles and splits the data.
I did not use it and it worked fine.

Srikanth Kumar
Graduate Student
The University of British Columbia
Sauder School of Business, Vancouver
Phone: +1 604 833 9309

**Comment by vincenzocoia** [comment]
FYI: I fixed the non-existing date on the assignment from "Wednesday" to "Monday".

**Comment by souyazdi** [comment]
Hello Vincenzo,

You said that "you can just remove observations where one of the predictors or response is missing."
So, you suggest that we cleanup the file first (i.e. to identify the rows with one or more null cells and save the data in new .csv file) before moving on to fitting the data?

**Comment by ChangHe591** [comment]
Hello Vincenzo,

3(b) For calculating the MSE of the test set, we will need to know the actual survival of people in the test set right? However, the survival is not given in test set data.

3(c)
![image](https://user-images.githubusercontent.com/36861381/37245217-bffd3c68-2449-11e8-88c4-b1da630909d8.png)
Do you mean" how does the training error curve differ from the test error curve"?
Also, same question here. How can we know the test set error rate if the actual survival is not given in test set data.

Thank you!
Chang

**Comment by ishakataria** [comment]
@Maggielena 
I am using similar code, but not assigning the shuffled values to a new array. Here is what I am doing, this might help, if not, happy to discuss offline. 

X = digits['data']
import numpy
numpy.random.shuffle(X)

**Comment by vincenzocoia** [comment]
> So, you suggest that we cleanup the file first (i.e. to identify the rows with one or more null cells and save the data in new .csv file) before moving on to fitting the data?

Almost. If an observation is missing in a column that we're not using for our analysis, then that shouldn't determine whether or not we remove that row. To do this, I recommend subsetting the data frame to the relevant columns, and then using the `na.omit` function (at least, this will work in R). There shouldn't be a need to re-write the data to csv, as you can just work internally 

> 3(b) For calculating the MSE of the test set, we will need to know the actual survival of people in the test set right? However, the survival is not given in test set data.

Aha, a little/big tidbit I failed to notice. Sorry about that. You'll just have to split the training data, as usual.

I'll answer the other questions soon!

Some hints for the assignment:

- 1(a)(1) `import numpy`, then you case use `numpy.linalg.norm`.
- 1(a)(3) To elaborate on the setup here, we're imagining that we have _both_ the image (pixels) and corresponding label for elements 1 and 2, but we only have the image (not the label) corresponding to element 0.
- 1(b)(2) See the code chunk below the text "Here's a plot of a random example:"? Copy and paste that, replace the `idx` variable with the element number that you want to plot, and you'll get a plot of the image.

**Comment by karinatkach** [comment]
Hi Vincenzo, 

If we want to use numpy.random.shuffle, do I understand correctly that I would have to use it on 2 arrays: X and Y? If so, how do I apply this method to 2 arrays simultaneously? 
For 2(a) I used the shuffle built in the sklearn.model_selection.train_test_split, but still I am wondering about numpy.random.shuffle. 

Thanks in advance for your response!

Karina

**Comment by goyaldipin1989** [comment]
Hi Vincenzo,

Regarding the shuffle of data for Question 2:

I understand that numpy.random.shuffle(X) function, only holds goods in case of a single dimensional array, however, our array X and Y, are multidimensional arrays.

So , how I do I shuffle multi-dimensional arrays?

**Comment by ishakataria** [comment]
I personally did numpy.random.shuffle(X) and numpy.random.shuffle(Y), and it worked fine for me. :)

**Comment by goyaldipin1989** [comment]
Hi isha 
numpy.random.shuffle function will work. 
However, our array is multi-dimensional, so not sure, if this is a right function!

**Comment by ishakataria** [comment]
Hi @vincenzocoia/MBAN friends,

When we use Loess, and perform scaling, are we supposed to:
1. Log transform fare and then replace missing values with the mean of the log of fare, or treat missing data first and then log transform?

2. Treat missing data and then split the data set into data set for male and data set for female? Or split the data first and then treat missing data?

3. To judge the appropriate bandwidth from the plot of training error curve/test error curve,  using the "validation set approach" for choosing hyperparameters, do we have to use the cross-validation approach or simply judge from the curves obtained initially? (with ref to Q2(b))

Thanks in advance.

**Comment by ishakataria** [comment]
@goyaldipin1989 , as per SciPy.org "This function only shuffles the array along the first axis of a multi-dimensional array. The order of sub-arrays is changed but their contents remains the same."

So I think this level of shuffling will serve our purpose to re-order data before splitting into training and test set. Hope that answers your concern.

**Comment by vincenzocoia** [comment]
> Do you mean" how does the training error curve differ from the test error curve"?

This is for 3c, and indeed I did mean "test error" curve.

-------------

Regarding questions about the random shuffling: it's actually better to use the `train_test_split` function from `sklearn.model_selection`. It does the randomization for you. This might help: (just replace 0.5 with the proportion you'd like to use for your test set)

```
X_train, X_test, y_train, y_test = train_test_split(X, Y, test_size=0.5)
```

--------------

@ishakataria :

> 1. When we use Loess, and perform scaling, are we supposed to:
>     - Log transform fare and then replace missing values with the mean of the log of fare, or treat missing data first and then log transform?
>     - Treat missing data and then split the data set into data set for male and data set for female? Or split the data first and then treat missing data?

You should first log-transform, then remove missing data -- only because I believe there are some `fare`s that are 0 (in which case, the log-transform will not exist). After removing data, it's then best to split into training and test (so that you have more control as to how much data is in each set).

> 2. (with ref to Q2(b))
> To judge the appropriate bandwidth from the plot of training error curve/test error curve,  using the "validation set approach" for choosing hyperparameters, do we have to use the cross-validation approach or simply judge from the curves obtained initially? 

You don't need to use cross validation to choose the bandwidth -- the validation set approach is sufficient.

**Comment by karinatkach** [comment]
Hi Vincenzo, 

1. na.omit does not actually remove the "-Inf" we get from log-transforming fares which are equal to 0. Can I just use ifelse and keep 0 in the log-transformed column if fare on the original scale is =0? 

2. If we were to split into female/male subsets after we split the data into the training/test sets, wouldn't we have some bias as we could end up having like 20% female test set and for male it could be 30%, so we'd have different test set sizes for male and female. Wouldn't it be better to first split into male/female subsets, and then split each subset into training and test sets?

3. And also could you please tell if we have some specific thresholds that would tell us if we need to do the scaling? Or is it just based on our judgement? 

Thank you in advance!

**Comment by vincenzocoia** [comment]
> 1. na.omit does not actually remove the "-Inf" we get from log-transforming fares which are equal to 0. Can I just use ifelse and keep 0 in the log-transformed column if fare on the original scale is =0?

I don't believe there are values of `fare` between 0 and 1, so I don't see any harm in keeping the fare as 0 even after the log transformation.

> 2. If we were to split into female/male subsets after we split the data into the training/test sets, wouldn't we have some bias as we could end up having like 20% female test set and for male it could be 30%, so we'd have different test set sizes for male and female. Wouldn't it be better to first split into male/female subsets, and then split each subset into training and test sets?

Sounds like a good idea to me.

> 3. And also could you please tell if we have some specific thresholds that would tell us if we need to do the scaling? Or is it just based on our judgement?

It's typically a good idea if your predictors are on different scales that can't be compared -- like centimeters and dollars. You can't really go wrong with scaling.

**Comment by riowxm** [comment]
When I'm doing Q3 in python, I found this article is very useful for data cleaning and deal with the missing data. 
Here is the link: https://www.shanelynn.ie/using-pandas-dataframe-creating-editing-viewing-data-in-python/

**Comment by vincenzocoia** [comment]
There were some questions about Exercise 3 that I'll answer here:

> 1. Scaling is necessary in the numeric predictors. But after doing log to fare, do we need to scale it further?

Yes

> 2. While fitting the loess, should i use all the 3 predictors together or one at a time?

Should be all together. After subsetting to `sex`, there should be only 2 predictors though.

> 3. Should we use the scaled age, sex and log of fare while fitting the loess?

Yes, but ideally, the scaled log fare instead of just log fare.

**Comment by howardwu8** [comment]
Hello professor:

When I use Python sklearn package, I come across several issues:

In the dataset, some of the data for 'age' and 'fare' are missing. As we learned from other classes, there are several ways to deal with missing data. While you might mention to some of my classmates that just delete the full rows of the missing data, I believe there are other ways such as full with 0 or mean value to handle the missing data. Please specify your requirement or it's up to us to choose.

I went ahead with deleting the full rows of the missing data, but then in the RadiusNeighborsRegressor part, after training the model from training set, and predict it using test data, the predictions including numerous 'NaN', and as bandwidth increases, the number of 'NaN' decreases. Now I face such situation: As bandwidth increases., the test set mse decreases and becomes lower than the mse of training set.. which does not make sense to me..

Please kindly advise what we should handle both the missing data from dataset, and missing data from prediction.

**Comment by vincenzocoia** [comment]
@howardwu8 

> While you might mention to some of my classmates that just delete the full rows of the missing data, I believe there are other ways such as full with 0 or mean value to handle the missing data. Please specify your requirement or it's up to us to choose.

Indeed there are several approaches to handling missing data. For the purpose of not complicating this assignment, the "complete case" method is sufficient -- that is, removing cases having a missing observation. I don't recommend mean imputation, as that tends to make the data unrealistically "central". If you do want to use a method besides the complete case, I'd recommend multiple imputation (you can use the `mice` package in R). 

> the predictions including numerous 'NaN', and as bandwidth increases, the number of 'NaN' decreases.

I'd expect this, if your bandwidth is too small, because there may not be any data in the subset. 

> Now I face such situation: As bandwidth increases., the test set mse decreases and becomes lower than the mse of training set.. which does not make sense to me..

A decreasing error for small bandwidths makes sense -- we anticipate a convex (U-shaped) curve. About the test error being lower than the training error, this happens sometimes, just due to randomness. If you choose different partitions of data into training and test sets, then you'd probably see that on average, the training error is smaller.

**Comment by nelsonwg** [comment]
Hi Professor, can you please give us a nudge on how to do 3d. 

I have no idea how to implement multi dimensional KNN. I've been trying to do it on R and I've been looking at these two functions:

https://stat.ethz.ch/R-manual/R-devel/library/class/html/knn.html
https://www.rdocumentation.org/packages/DMwR/versions/0.4.1/topics/kNN

The first one is from the caret package but I have no idea what is CL and how to specify it properly. The second one is from the DMwR package and I have no idea how to specify the "form" properly. It looks like I'm supposed to put a formula like Survived ~ Age.Scaled + Fare.Scaled, but I keep getting an error saying "Error in colMeans(x, na.rm = TRUE) : 'x' must be numeric".

For the DMwR package I have tried the following, both yielding the same error:
knn.f <- kNN(Survived ~ ., train.male, test.male, k = 5)
knn.f <- kNN(Survived ~ Age.Scaled + Fare.Scald, train.male, test.male, k= 5)

Finally, I tried:
knn.f <- kNN(train.male$Survived ~ train.male$Age.Scaled + train.male$Age., train.male, test.male, k = 5)
but that gives me:
Error in knn(train[, -tgtCol], test[, -tgtCol], train[, tgtCol], ...) : 
  'train' and 'class' have different lengths

**Comment by vnirwan** [comment]
Hi Vincenzo,

Q3(b) :Fit a loess model to the training data for various values of the bandwidth parameter.
Do we have to convert Survived to a categorical variable before making the model?

**Comment by mooahmed** [comment]
@nelsonwg 
Have you tried looking at the values that you have in train.male? 
Maybe there is something wrong with those values ;)

**Comment by mooahmed** [comment]
@vnirwan 
No, you can read the data and fir the model directly.

**Comment by vincenzocoia** [comment]
This might be handy: when using the `mean()` function in R, including `na.rm=TRUE` will ignore missing values. Otherwise, if even one `NA` is present in the vector, an NA is returned as well.

**Comment by vincenzocoia** [comment]
As requested, I'll provide a hint for 3(d). It's probably enough to elaborate on the documentation of the `knn` function from the package `class`. 

- The `train` argument should be a data frame containing one column for each predictor, and _no other columns_ (so, the response should not appear here). Each row should represent an observation on the training data.
- The `test` argument is the same as `train`, but it will make predictions on this set.
- The cl argument is a vector of the response -- so, the `Survived` column in your data.
- The `k` argument is k in kNN. 
- The output is a vector of predictions on the data in `test`.

Just put the training data in both the `train` and `test` argument to obtain predictions on the training data.

**Comment by Jimyie** [comment]
Hi professor, 

For Q3d, do you mean that we need to use k-NN to get the prediction of its survival probability and then classify the probability to 1 or 0 based on the threshold 0.5? Or you mean that we just need to use the k-NN to classify each person to get its survival status whether it is 1 or 0?

I am wondering that if you mean we should classify each person by the former method, how can I get the result of its survival  probability by k-NN? Because k-NN is often used to classify each observation directly.

Thank you!

**Comment by mooahmed** [comment]
Hi @Jimyie,

Depending on the implementation that you are using for k-NN, you get either the probability or the predicted class.
As you mentioned, if your k-nn code returns the probability, then use a threshold of 0.5 to get the predicted class.
